File: app/src/main/java/com/example/moneygement/controller/DispLedgerGraphActivity.java
package com.example.moneygement.controller;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.os.DropBoxManager;
import android.view.View;
import android.widget.Button;

import com.example.moneygement.R;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.utils.ColorTemplate;

import java.util.ArrayList;
import java.util.List;

public class DispLedgerGraphActivity extends AppCompatActivity {

    float rainfall[] = {98.8f, 123.8f, 34.6f, 43.9f, 69.4f};
    String monthNames[] = {"食費", "趣味", "光熱費", "交通費", "日用品"};




    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_disp_ledger_graph);

        setupPieChart();
    }

    private void setupPieChart() {
        //PieEntriesのリストを作成する:
        List<PieEntry> pieEntries = new ArrayList<>();
        for (int i = 0; i < rainfall.length; i++) {
            pieEntries.add(new PieEntry(rainfall[i], monthNames[i]));
        }
//
        PieDataSet dataSet = new PieDataSet(pieEntries, "Rainfall for Vancouver");
        dataSet.setColors(ColorTemplate.COLORFUL_COLORS);
        PieData data = new PieData(dataSet);
//
        //PieChartを取得する:
        PieChart piechart1 = (PieChart) findViewById(R.id.chart1);
        piechart1.setData(data);
        piechart1.invalidate();

        PieChart piechart2 = (PieChart) findViewById(R.id.chart2);
        piechart2.setData(data);
        piechart2.invalidate();

        Button cancel = (Button)findViewById(R.id.cancel);
        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(DispLedgerGraphActivity.this,MainActivity.class);
                startActivity(intent);
            }
        });
    }
}

File: app/src/main/java/com/example/moneygement/controller/DispLedgerGraphActivity.kt
package com.example.moneygement.controller

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.example.moneygement.R
import com.example.moneygement.databinding.ActivityDispLedgerGraphBinding
import com.example.moneygement.viewmodel.LedgerGraphViewModel
import com.example.moneygement.viewmodel.LedgerViewModel
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.utils.ColorTemplate
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.util.*

class DispLedgerGraphActivity : AppCompatActivity() {
    private lateinit var binding: ActivityDispLedgerGraphBinding
    var ledgerId = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_disp_ledger_graph)

        var intent = intent
        ledgerId = intent.getIntExtra("ledgerId", 0)


//      viewModelを結び付けているが、binding的役割はない・・・泣
        val viewModel: LedgerGraphViewModel by viewModels()
        binding = DataBindingUtil.setContentView(
                this,
                R.layout.activity_disp_ledger_graph
        )
        binding.lifecycleOwner = this@DispLedgerGraphActivity
        binding.vm = viewModel

        var incomeGraph = findViewById<View>(R.id.chart2) as PieChart
        var expenseGraph = findViewById<View>(R.id.chart1) as PieChart

//      非同期でviewModelのメソッドを呼びデータが取得され次第、画面に適用させる。
        if(ledgerId == 0){
            var intent = Intent(this@DispLedgerGraphActivity, MainActivity::class.java)
            startActivity(intent)
        }

        GlobalScope.launch {
            viewModel.getLedgeGraphData(ledgerId)
            incomeGraph.data = PieData(viewModel.incomeGraph)
            expenseGraph.data = PieData(viewModel.expenseGraph)
            incomeGraph.invalidate()
            expenseGraph.invalidate()
        }

        val cancel = findViewById<View>(R.id.cancel) as Button
        cancel.setOnClickListener {
            onBackPressed()
        }
    }


//    private fun setupPieChart() {
//        //PieEntriesのリストを作成する:
//        val pieEntries: MutableList<PieEntry> = ArrayList()
//        for (i in incomesTotalMoney.indices) {
//            pieEntries.add(PieEntry(incomesTotalMoney[i], category[i]))
//        }
//        //
//        val dataSet = PieDataSet(pieEntries, "Rainfall for Vancouver")
//        dataSet.setColors(*ColorTemplate.COLORFUL_COLORS)
//        val data = PieData(dataSet)
//        //
//        //PieChartを取得する:
//        val piechart1 = findViewById<View>(R.id.chart1) as PieChart
//        piechart1.data = data
//        piechart1.invalidate()
//        val piechart2 = findViewById<View>(R.id.chart2) as PieChart
//        piechart2.data = data
//        piechart2.invalidate()
}